src/System.Data.Common/src/System/Data/Common/DbProviderFactories.cs

```diff
@@ -0,0 +1,187 @@
+                    }
+                    throw ADP.Argument(SR.Format(SR.ADP_DbProviderFactories_FactoryNotLoadable, assemblyQualifiedName));
+                }
+            }
+            throw ADP.Argument(SR.ADP_DbProviderFactories_NoAssemblyQualifiedName);
+        }
+
+        public static DataTable GetFactoryClasses()
+        {
+            try
+            {
+                _providerTableLock.EnterReadLock();
+                return _providerTable == null ? GetInitialProviderTable() : _providerTable.Copy();
+            }
+            finally
+            {
+                _providerTableLock.ExitReadLock();
+            }
+        }
+
+        public static DbProviderFactory GetFactory(DbConnection connection)
+        {
+            ADP.CheckArgumentNull(connection, nameof(connection));
+            return connection.ProviderFactory;
+        }
+
+        public static void ConfigureFactory(Type providerFactoryClass, string providerInvariantName)
+        {
+            ConfigureFactory(providerFactoryClass, providerInvariantName, string.Empty, string.Empty);
+        }
+        
+        public static void ConfigureFactory(Type providerFactoryClass, string providerInvariantName, string name, string description)
+        {
+            ADP.CheckArgumentNull(providerFactoryClass, nameof(providerFactoryClass));
+            ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
+
+            if (!providerFactoryClass.IsSubclassOf(typeof(DbProviderFactory)))
+            {
+                throw ADP.Argument(SR.Format(SR.ADP_DbProviderFactories_NotAFactoryType, providerFactoryClass.FullName));
+            }
+            RegisterFactoryInTable(providerFactoryClass, providerInvariantName, name, description);
+        }
+        
+        public static void ConfigureFactory(DbConnection connection, string providerInvariantName)
+        {
+            ConfigureFactory(connection, providerInvariantName, string.Empty, string.Empty);
+        }
+        
+        public static void ConfigureFactory(DbConnection connection, string providerInvariantName, string name, string description)
+        {
+            ADP.CheckArgumentNull(connection, nameof(connection));
+            ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
+
+            DbProviderFactory factoryInstance = GetFactory(connection);
+            if (factoryInstance == null)
+            {
+                throw ADP.Argument(SR.ADP_DbProviderFactories_NoFactorySuppliedThroughDbConnection);
+            }
+            RegisterFactoryInTable(factoryInstance.GetType(), providerInvariantName, name, description);
+        }
+        
+        private static void RegisterFactoryInTable(Type factoryType, string providerInvariantName, string name, string description)
+        {
+            ADP.CheckArgumentNull(factoryType, nameof(factoryType));
```

Comment:
@FransBouma I am a bit lost. It seems the fallback to namespace is currently never used. Doesn't that mean it can just be removed?

It this was one of the repos over at https://github.com/aspnet I would feel more confident suggesting the private method doesn't need to have the check if you are sure it is checked in all the public entry points, but there is some variability, e.g. you could choose to have a `Debug.Assert()` instead. So I would start with @karelz for corefx guidance and @saurabh500 for current practices.

And while we are on it, can we switch to `var` everything that can be switched to `var` in this code?